llvm/utils/TableGen/GlobalISelEmitter.cpp

```diff
@@ -1412,15 +1412,18 @@ Expected<BuildMIAction &> GlobalISelEmitter::createAndImportInstructionRenderer(
   action_iterator InsertPt = InsertPtOrError.get();
   BuildMIAction &DstMIBuilder = *static_cast<BuildMIAction *>(InsertPt->get());
 
-  for (auto PhysInput : InsnMatcher.getPhysRegInputs()) {
-    InsertPt = M.insertAction<BuildMIAction>(
-        InsertPt, M.allocateOutputInsnID(),
-        &Target.getInstruction(RK.getDef("COPY")));
-    BuildMIAction &CopyToPhysRegMIBuilder =
-        *static_cast<BuildMIAction *>(InsertPt->get());
-    CopyToPhysRegMIBuilder.addRenderer<AddRegisterRenderer>(
-        Target, PhysInput.first, true);
-    CopyToPhysRegMIBuilder.addRenderer<CopyPhysRegRenderer>(PhysInput.first);
+  for (auto PhysOp : M.physoperands()) {
+    auto &OpInsnMatcher = PhysOp.second->getInstructionMatcher();
+    for (auto PhysInput : OpInsnMatcher.getPhysRegInputs()) {
```

Comment:
If there are more than one phys reg inputs, this loop nest will add them several times.
It looks like this map iterated in the inner loop is redundant, `physregoperands` returns all physreg inputs.

Reply:
Interesting, if we have a pattern like 
```
def ADD_PHYS : I<(outs GPR32:$dst), (ins),
    [(set GPR32:$dst, (add SPECIAL, SPECIAL))]> {
  let Uses = [SPECIAL];
}
```
We see two copies without this change as well
```
      // (add:{ *:[i32] } SPECIAL:{ *:[i32] }, SPECIAL:{ *:[i32] })  =>  (ADD_PHYS:{ *:[i32] })
      GIR_BuildMI, /*InsnID*/2, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/2, GIMT_Encode2(MyTarget::SPECIAL), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/2, /*OldInsnID*/0, /*OpIdx*/1, // SPECIAL
      GIR_BuildMI, /*InsnID*/1, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/1, GIMT_Encode2(MyTarget::SPECIAL), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/1, /*OldInsnID*/0, /*OpIdx*/1, // SPECIAL
      GIR_BuildRootMI, /*Opcode*/GIMT_Encode2(MyTarget::ADD_PHYS),
```
At the same time for the pattern like
```
def ADD_PHYS : I<(outs GPR32:$dst), (ins ),
    [(set GPR32:$dst, (add SPECIAL, SPECIAL_2))]> {
  let Uses = [SPECIAL, SPECIAL_2];
}
```
We need this loop because two copies are required
```
      // (add:{ *:[i32] } SPECIAL:{ *:[i32] }, SPECIAL_2:{ *:[i32] })  =>  (ADD_PHYS:{ *:[i32] })
      GIR_BuildMI, /*InsnID*/2, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/2, GIMT_Encode2(MyTarget::SPECIAL_2), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/2, /*OldInsnID*/0, /*OpIdx*/2, // SPECIAL_2
      GIR_BuildMI, /*InsnID*/1, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/1, GIMT_Encode2(MyTarget::SPECIAL), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/1, /*OldInsnID*/0, /*OpIdx*/1, // SPECIAL
      GIR_BuildRootMI, /*Opcode*/GIMT_Encode2(MyTarget::ADD_PHYS),
```
So should we have initially tracked the uniqueness of physical registers?

Reply:
> So should we have initially tracked the uniqueness of physical registers?

Yes, probably. But, a) support has been limited to simple patterns and b) SelectionDAG doesn't seem to unique them either. Uniqueness wasn't my point, please see below.

> We need this loop because two copies are required

Do we need both loops? I didn't try it but it looks like only the outer loop is required (?). The inner loop iterates over the same set of registers as the outer loop, so it looks like we will end up with four registers added instead of two. Am I missing something here?

Reply:
Here is a test
```
def MULM_PHYS : I<(outs), (ins),
    [(st GPR32:$src0, (mul R0, SPECIAL))]> {
  let Uses = [R0, SPECIAL];
}
```

With this patch this adds each of R0 and SPECIAL twice:

```
      // (st GPR32:{ *:[i32] }:$src0, (mul:{ *:[i32] } R0:{ *:[i32] }, SPECIAL:{ *:[i32] }))  =>  (MULM_PHYS)
      GIR_BuildMI, /*InsnID*/4, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/4, GIMT_Encode2(MyTarget::SPECIAL), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/4, /*OldInsnID*/1, /*OpIdx*/2, // SPECIAL
      GIR_BuildMI, /*InsnID*/3, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/3, GIMT_Encode2(MyTarget::R0), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/3, /*OldInsnID*/1, /*OpIdx*/1, // R0
      GIR_BuildMI, /*InsnID*/2, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/2, GIMT_Encode2(MyTarget::SPECIAL), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/2, /*OldInsnID*/1, /*OpIdx*/2, // SPECIAL
      GIR_BuildMI, /*InsnID*/1, /*Opcode*/GIMT_Encode2(TargetOpcode::COPY),
      GIR_AddRegister, /*InsnID*/1, GIMT_Encode2(MyTarget::R0), /*AddRegisterRegFlags*/GIMT_Encode2(RegState::Define),
      GIR_Copy, /*NewInsnID*/1, /*OldInsnID*/1, /*OpIdx*/1, // R0
      GIR_BuildRootMI, /*Opcode*/GIMT_Encode2(MyTarget::MULM_PHYS),
      GIR_MergeMemOperands, /*InsnID*/0, /*NumInsns*/2, /*MergeInsnID's*/0, 1,
      GIR_RootConstrainSelectedInstOperands,
      // GIR_Coverage, 0,
      GIR_EraseRootFromParent_Done,
```

Reply:
@s-barannikov yes, thank you! I'll add this as a test as well.
I assumed that InsnMatcher is a must but we can just take keys from `PhysRegOperands` as only they are needed. And the duplication problem is solved completely as `PhysRegOperands` is a `DenseMap`. Only named phys registers are a problem but it looks like they are not supported yet. The test for them is commented out.

Reply:
I think SelectionDAG doesn't support named physical registers either, so this shouldn't be an issue (in the end, we're trying to catch up with SelectionDAG, not overtake it).